src/cache/base.py

```diff
@@ -0,0 +1,102 @@
+from functools import wraps
+from typing import Any, Callable
+from sys import getsizeof
+import pickle
+import time
+
+
+class BaseCache:
+    def __init__(self) -> None:
+        # these specific elements can be change subClasses
+        self.dumper = pickle.dumps
+        self.loader = pickle.loads
+
+    def _log_cache_failed(self, key: str, logger: Any) -> None:
+        """Just some logging"""
+        logger.debug(f">> cache failed on custom_condition for {key=}")
+
+    def _log_cache_hit(self, key: str, logger: Any) -> None:
+        """Just some logging"""
+        logger.debug(f">> cache hit for {key=}")
+
+    def _log_cache_missed(self, key: str, logger: Any) -> None:
+        """Just some logging"""
+        logger.debug(f">> cache missed for {key=}")
+
+    def _size_of(self, value: Any) -> int:
+        """To return the memory size of a value"""
+        return getsizeof(value)
+
+    def _build_key(self, *args: Any, **kwargs: Any) -> str:
+        """The key builder for our cache system"""
+        return f"{args}-{kwargs}"
+
+    def _build_value(self, value: Any) -> bytes:
+        """
+        We need to have an uniform way to build the value and how it will be stored
+        as bytes for perfs requirements
+        """
+        return self.dumper(
+            {
+                "value": value,
+                "at": time.time(),
+                "size_of": self._size_of(value),
+            }
+        )
+
+    def _load_value(self, value: Any) -> Any:
+        return self.loader(value)
+
+    def get(
+        self, key: str, default: Any | None = None, custom_condition: bool = True
+    ) -> Any:
+        """To get something from the cache based on many conditions"""
+        ...
+
+    def set(
+        self, key: str, value: Any | None = None, custom_condition: bool = True
```

Comment:
What would be the usage of custom_condition?

Reply:
we can have situation where we want to pass the cache execution based on the result of something, this help us not implementing, I can of agree it's more estetical than really indispensable !

```python
self.set(key="this-key", value="this-value", is_something_working(True, param) == [0])
```

rather than 
 ```python
if is_something_working(True, param) == [0]:
	self.set(key="this-key", value="this-value")
```

I was also thinking of an "after cache callback" with its args that we can pass to set/get... but didn't implement it because it was looking too complex for nothing really important